### ISSUE TITLE
Emit event upon setting upgrade consensus state

### ISSUE THREAD

**Issue by colin-axner** [cosmos/ibc-go#270]
## Summary

When we set the upgrade consensus state, we should emit an event so relayers know that they can now upgrade the counterparty client. Unsure if this needs to be backported since I don't know if hermes can listen to events in begin/end blocker without the next tendermint version

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged/assigned

**Comment by technicallyty** [comment]
i think this can be closed, https://github.com/cosmos/ibc-go/blob/main/modules/core/02-client/keeper/client.go#L191-L198 ?

**Comment by colin-axner** [comment]
No, we need the entire consensus state to be emitted when it is being set before a chain [upgrades](https://github.com/cosmos/ibc-go/blob/823ef675bbf38a48faa3ba670fb6b8228d58b9fd/modules/core/02-client/abci.go#L22). The event emitted is on the counterparty chain (which didn't upgrade). It is emitting an event for the processing of a message which conveys proof of the counterparty upgrade.

Chain A upgrades (setting client/consensus states)
Chain B processed MsgUpgradeClient which shows proof of chain A upgrading

**Comment by adizere** [comment]
This issue is a dependency for https://github.com/informalsystems/ibc-rs/issues/1148.



> Unsure if this needs to be backported since I don't know if hermes can listen to events in begin/end blocker without the next tendermint version

Can confirm that Hermes at the moment can listen & act on begin/end block events (https://github.com/informalsystems/ibc-rs/issues/1793).

**Comment by colin-axner** [comment]
The event function used should something like look like:
```go
func EmitUpgradeChainEvent(ctx sdk.Context, height int64) {
	ctx.EventManager().EmitEvents(sdk.Events{
		sdk.NewEvent(
			types.EventTypeUpgradeChain,
			sdk.NewAttribute(types.AttributeKeyPlanHeight, height), 
	                sdk.NewAttribute(types.AttributeKeyUpgradeStore, upgradetypes.StoreKey), // which store to query proof of consensus state from
		),
	})
}
```

Is any other information needed @adizere? Will relayers know which counterparty clients belong to this chain that is upgrading?

**Comment by adizere** [comment]
This looks good at first glance.

Is there a way we can actually test this in practice and reproduce the event emission? I'm asking because it's difficult to say a firm yes/no whether further info would be necessary unless we run through the whole exercise end-to-end. Would be glad to use simapp to test this new event, or is there another way? Any instructions would be appreciated!

> Will relayers know which counterparty clients belong to this chain that is upgrading?
I think relayers should know yes; they should have support for tracking the relevant counterparty clients that need to be upgraded.

**Comment by crodriguezvega** [comment]
>Is there a way we can actually test this in practice and reproduce the event emission? I'm asking because it's difficult to say a firm yes/no whether further info would be necessary unless we run through the whole exercise end-to-end. Would be glad to use simapp to test this new event, or is there another way? Any instructions would be appreciated!

Yes, we can make the changes in a branch and you could build simapp from that branch to test.

**Comment by catShaark** [comment]
Can we test using unit test instead of running the whole thing e2e.

**Comment by crodriguezvega** [comment]
> Can we test this using test unit instead of running the whole thing e2e ?

It's a bit overkill to assert in the tests that the right values are emitted in the events (i.e. you need to loop through a `[]abci.Event` and for each event loop through again a `[]abci.EventAttribute` to find the attributes you want to assert). So I think in this case it's not really worth it...

**Comment by gjermundgaraba** [comment]
This issue was mistakenly reopened during some integration work, so closing again now.
